File: problems/05_binary_search/0981_time_based_key_value_store/solution.swift
// Problem: 0981. Time Based Key-Value Store
// Difficulty: Medium

import Foundation

class TimeMap {
    private var store: [String: [(Int, String)]] = [:]

    func set(_ key: String, _ value: String, _ timestamp: Int) {
        // TODO: implement
    }

    func get(_ key: String, _ timestamp: Int) -> String {
        // TODO: implement
        return ""
    }
}

func checkStr(_ a: String, _ e: String, _ msg: String) {
    if a != e { fatalError("FAIL \(msg): \"\(a)\" vs \"\(e)\"") }
}

// MARK: - Tests
// let s = Solution()

// Test 1
// let result = s.solve(...)

print("All tests passed!")

File: problems/05_binary_search/0981_time_based_key_value_store/solution.py
"""
Problem: 0981. Time Based Key-Value Store
Difficulty: Medium
URL: https://leetcode.com/problems/time-based-key-value-store/

解法演進：
1) 暴力 get 線性 O(n)。
2) 每 key 一個升序陣列 + 二分：set O(1), get O(log n)。

最佳解核心思路：
- set 時間戳嚴格遞增 -> 陣列自然有序，可直接二分找 <= timestamp 的最右位置。

複雜度：set O(1)，get O(log n)。
"""

class TimeMap:
    def __init__(self) -> None:
        pass

    def set(self, key: str, value: str, timestamp: int) -> None:
        pass

    def get(self, key: str, timestamp: int) -> str:
        pass


if __name__ == "__main__":
    timeMap = TimeMap()

    # Test 1
    # result = timeMap.solve(...)
    # assert result == expected

    print("All tests passed!")
